File: 57-InsertInterval/57-InsertInterval.java
// Last updated: 8/16/2026, 9:53:01 PM
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        /* 
        1. add all intervals i that i_end < newInterval_start, i++
        2. for overlapping intervals, merge:
            merge_start = min(i_start, newInterval_start)
            merge_end = max(i_end, newInterval_end)
        3. for remianing intervals, add to results
        */
        List<int[]> res = new ArrayList<>();
        int newS = newInterval[0], newE = newInterval[1];
        int n = intervals.length;
        int i = 0;
        while(i < n && intervals[i][1] < newS) {
            res.add(Arrays.copyOf(intervals[i], 2));
            i ++;
        }
        int[] merged = Arrays.copyOf(newInterval, 2);
        while(i < n && intervals[i][0] <= newE) {
            // find ALL overlappings
            merged[0] = Math.min(merged[0], intervals[i][0]);
            merged[1] = Math.max(merged[1], intervals[i][1]);
            i ++;
            System.out.println(merged[0]);
            System.out.println(merged[1]);
        }
        res.add(merged); // add merged interval
        while(i < n) {
            res.add(Arrays.copyOf(intervals[i], 2));
            i ++;
        }
        // convert List<> to int[]
        int[][] res_ = new int[res.size()][2];
        for(int p = 0; p < res.size(); p ++) {
            res_[p][0] = res.get(p)[0];
            res_[p][1] = res.get(p)[1];
        }
        return res_;
    }
}

File: 57-InsertInterval/57-InsertInterval.py
# Last updated: 9/1/2026, 8:28:58 PM
1class Solution:
2    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
3        # [s1, e1] for every interval, [s2,e2]: new_interval
4        # leftside of new:          e1 < s2
5        # overlap, need combine:    e1 >= s2 & s1 <= e2
6        # rightside of new:         s1 > e2
7
8        i, n = 0, len(intervals)
9        s2, e2 = newInterval[0], newInterval[1]
10
11        res = []
12
13        # leftside, copy directly
14        while i < n and intervals[i][1] < s2:
15            res.append(intervals[i])
16            i += 1
17        
18        # update [s2,e2] when overlapping
19        while i < n and intervals[i][0] <= e2:
20            s2 = min(intervals[i][0], s2)
21            e2 = max(intervals[i][1], e2)
22            i += 1
23        res.append([s2, e2])
24
25        # rightside, copy directly
26        while i < n:
27            res.append(intervals[i])
28            i += 1
29        
30        return res
31
32            
33
34